### ISSUE TITLE
media-feature-no-missing-punctuation should ignore non-standard syntax

### ISSUE THREAD

**Issue by niksy** [stylelint/stylelint#1648]
> Description

When using `media-feature-no-missing-punctuation`, warning is reported when using parenthesis, either as separator or as part of function like `calc()`.

> How to reproduce

`index.scss`

```scss
@media screen and (max-width:42px) {

}

@media screen and (max-width:calc(100% - 42px)) {

}

@media screen and (max-width:($var - 42px)) {

}
```

`.stylelintrc`

```json
{
	"rules": {
		"media-feature-no-missing-punctuation": true
	}
}
```

CLI output when using `stylelint index.scss --syntax scss`

```sh
index.scss
 5:19  ✖  Unexpected missing punctuation   media-feature-no-missing-punctuation
 9:19  ✖  Unexpected missing punctuation   media-feature-no-missing-punctuation
```

---

All provided media queries are valid, except maybe last on when used outside SCSS file.

**Comment by evilebottnawi** [comment]
@niksy Thanks for reporting.

**Comment by evilebottnawi** [comment]
Confirm

**Comment by niksy** [comment]
Additional information says that [media query values should only accept simple values](https://www.w3.org/TR/css3-mediaqueries/#media1):

> Properties may accept more complex values, e.g., calculations that involve several other values. Media features only accept single values: one keyword, one number, or a number with a unit identifier. (The only exceptions are the ‘aspect-ratio’ and ‘device-aspect-ratio’ media features.)

But, SCSS files can have functions inside media queries, so I don’t know if this should be part of the core or let postcss-scss deal with this.

**Comment by davidtheclark** [comment]
We should ignore the media feature if it contains non-standard syntax.

**Comment by evilebottnawi** [comment]
@davidtheclark `@media screen and (max-width:calc(100% - 42px)) {}` also bugs, but it is standard syntax

**Comment by davidtheclark** [comment]
My understanding from @niksy's quotation above is that it's *not* standard syntax because you're not supposed to use a function like `calc()` inside the media feature.

**Comment by evilebottnawi** [comment]
@davidtheclark yes, you are right https://www.w3.org/TR/css3-mediaqueries/#media1

**Comment by niksy** [comment]
I guess this won’t be resolved then?

**Comment by evilebottnawi** [comment]
@niksy It will be resolved, perhaps tomorrow will be a time or you can resolve this self

**Comment by jeddy3** [comment]
> I guess this won’t be resolved then?

It'll be resolved by testing if the media query conforms to standard syntax. If it doesn't this rule will simply *ignore the media query outright*, which will nullify the two warnings for you :)

**Comment by evilebottnawi** [comment]
Ref: https://github.com/dryoma/postcss-media-query-parser/issues/1 can help resolve this issue, without check is standard, because parser undestand non-standard syntax

**Comment by jeddy3** [comment]
> without check is standard, because parser undestand non-standard syntax

I feel we should still create a `isStandardSyntaxMediaFeature` util and bail early. That behaviour would be consistent with our approach to non-standard syntax else where. It is a clear and unambiguous stance on how we *distance* the linter from non-standard syntax. I think relying on `postcss-media-query-parser` heuristics might unduly draw us closer to the idiosyncrasies (and possible bugs) of parsing non-standard syntax.

**Comment by evilebottnawi** [comment]
@jeddy3 agree, i found resolve my issue for simplifying rules and also create isStandardSyntaxMediaFeature

**Comment by jeddy3** [comment]
> i wait for resolve my issue for simplifying rules

Yeap, agreed. The use of `postcss-media-query-parser` will simplify this rule and use of `isStandardSyntaxMediaFeature` will protect it from non-standard syntax. I'm looking forward to seeing the improvements that these two additions bring :)
